src/components/Personalization/turbine/index.js

```diff
@@ -0,0 +1,151 @@
+  };
+
+  const logActionError = (action, rule, e) => {
+    logger.error(getErrorMessage(action, rule, e.message, e.stack));
+  };
+
+  const logConditionError = (condition, rule, e) => {
+    logger.error(getErrorMessage(condition, rule, e.message, e.stack));
+  };
+
+  const logConditionNotMet = (condition, rule) => {
+    const conditionDisplayName = getModuleDisplayNameByRuleComponent(condition);
+
+    logger.log(
+      `Condition ${conditionDisplayName} for rule ${rule.name} not met.`
+    );
+  };
+
+  const logRuleCompleted = rule => {
+    logger.log(`Rule ${rule.name} fired.`);
+  };
+
+  const runActions = (rule, syntheticEvent) => {
+    if (!rule.actions) {
+      return;
+    }
+
+    let action;
+
+    for (let i = 0; i < rule.actions.length; i += 1) {
+      action = rule.actions[i];
+
+      try {
+        executeDelegateModule(action, syntheticEvent, [syntheticEvent]);
+      } catch (e) {
+        logActionError(action, rule, e);
+        return;
+      }
+    }
+
+    logRuleCompleted(rule);
+  };
+
+  const checkConditions = (rule, syntheticEvent) => {
+    if (!rule.conditions) {
+      return;
+    }
+
+    let condition;
+
+    for (let i = 0; i < rule.conditions.length; i += 1) {
+      condition = rule.conditions[i];
+
+      try {
+        const result = executeDelegateModule(condition, syntheticEvent, [
+          syntheticEvent
+        ]);
+
+        if (!isConditionMet(condition, result)) {
+          logConditionNotMet(condition, rule);
+          return;
+        }
+      } catch (e) {
+        logConditionError(condition, rule, e);
+        return;
+      }
+    }
+
+    runActions(rule, syntheticEvent);
+  };
+
+  const initEventModule = ruleEventPair => {
+    const { rule, event } = ruleEventPair;
+    event.settings = event.settings || {};
+
+    let moduleName;
+    let extensionName;
+
+    try {
+      moduleName = moduleProvider.getModuleDefinition(event.modulePath).name;
+      extensionName = moduleProvider.getModuleExtensionName(event.modulePath);
+
+      const syntheticEventMeta = {
+        $type: `${extensionName}.${moduleName}`,
+        $rule: {
+          id: rule.id,
+          name: rule.name
+        }
+      };
+
+      const trigger = syntheticEvent => {
+        if (!eventModulesInitialized) {
+          triggerCallQueue.push(trigger.bind(null, syntheticEvent));
+          return;
+        }
+
+        const normalizedSyntheticEvent = normalizeSyntheticEvent(
+          syntheticEventMeta,
+          syntheticEvent
+        );
+
+        checkConditions(rule, normalizedSyntheticEvent);
+      };
+
+      executeDelegateModule(event, null, [trigger]);
+    } catch (e) {
+      logger.error(getErrorMessage(event, rule, e.message, e.stack));
+    }
+  };
+
+  buildRuleExecutionOrder(rules).forEach(initEventModule);
+  eventModulesInitialized = true;
+  triggerCallQueue.forEach(triggerCall => triggerCall());
+  triggerCallQueue = null;
+
+  return lastPromiseInQueue;
```

Comment:
You should be able to remove `lastPromiseInQueue`.

Reply:
Done.